# Projet télédétection
Audrey Zimmer

## 1. Import des librairies

In [11]:
# Librairies python
import sys
sys.path.append('..')
import os
from sklearn.model_selection import train_test_split
import numpy as np
import plotly.express as px
from sklearn import tree
from sklearn.metrics import confusion_matrix, classification_report, \
    accuracy_score
import geopandas as gpd
from osgeo import ogr
from osgeo import gdal
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import RepeatedStratifiedKFold, KFold


# Librairies personnelles
from libsigma import classification as cla
from libsigma import read_and_write as rw
from libsigma import plots


## Création des dossiers

In [12]:
# Chemin de base (absolu)
my_folder = "/home/onyxia/work/Projet_teledetection"

# Chemins des dossiers à créer (absolus)
results_path = os.path.join(my_folder, "../results")
figure_path = os.path.join(my_folder, "../results/figure")
img_path = os.path.join(my_folder, "img")

# Création des dossiers
os.makedirs(results_path, exist_ok=True)
os.makedirs(figure_path, exist_ok=True)
os.makedirs(img_path, exist_ok=True)

print("Dossiers results, figure et img créés avec succès !")


Dossiers results, figure et img créés avec succès !


## Formatage des données

In [ ]:
b03 = '/home/onyxia/work/data/projet_eval/bretagne_23-24_B03.tif'
b03_data_set = gdal.Open(b03, gdal.GA_Update)
type(data_set)

b05 = '/home/onyxia/work/data/projet_eval/bretagne_23-24_B05.tif'
b05_data_set = gdal.Open(b05, gdal.GA_Update)
type(data_set)

osgeo.gdal.Dataset

In [16]:
print('Nombre de colonnes b03 :', data_set.RasterXSize)
print('Nombre de lignes b03 :', data_set.RasterYSize)
print('Nombre de bandes b03 :', data_set.RasterCount)

print('Nombre de colonnes b05 :', data_set.RasterXSize)
print('Nombre de lignes b05 :', data_set.RasterYSize)
print('Nombre de bandes b05 :', data_set.RasterCount)

Nombre de colonnes b03 : 1533
Nombre de lignes b03 : 612
Nombre de bandes b03 : 5
Nombre de colonnes b05 : 1533
Nombre de lignes b05 : 612
Nombre de bandes b05 : 5


Les bandes 03 et 05 qui sont utilisées pour la suite de l'étude contiennent le même nombre de colonnes (1533), le même nombre de lignes (612) et le même nombre de bandes (5).

In [13]:
# Liste des bandes
bandes = ['B02', 'B03', 'B04', 'B05', 'B06', 'B07', 'B08', 'B8A', 'B11', 'B12']

# Dictionnaire pour stocker les datasets
datasets = {}

# Vérification des dimensions
for bande in bandes:
    chemin = f'/home/onyxia/work/data/projet_eval/bretagne_23-24_{bande}.tif'
    datasets[bande] = gdal.Open(chemin, gdal.GA_ReadOnly)
    print(f"Bande {bande}:")
    print(f"  - Colonnes: {datasets[bande].RasterXSize}")
    print(f"  - Lignes: {datasets[bande].RasterYSize}")
    print(f"  - Nombre de bandes: {datasets[bande].RasterCount}\n")


Bande B02:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B03:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B04:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B05:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B06:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B07:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B08:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B8A:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B11:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5

Bande B12:
  - Colonnes: 1533
  - Lignes: 612
  - Nombre de bandes: 5



/opt/python/lib/python3.13/site-packages/osgeo/gdal.py:330: FutureWarning:

Neither gdal.UseExceptions() nor gdal.DontUseExceptions() has been explicitly called. In GDAL 4.0, exceptions will be enabled by default.



Toutes les bandes obtenues ont la même dimension : 
- 1533 colonnes
- 612 lignes
- 5 bandes

In [14]:
# Chargement des datasets
for bande in bandes:
    chemin = f'/home/onyxia/work/data/projet_eval/bretagne_23-24_{bande}.tif'
    datasets[bande] = gdal.Open(chemin, gdal.GA_ReadOnly)

# Nombre de dates (5)
nb_dates = datasets['B02'].RasterCount

# Dimensions spatiales
cols = datasets['B02'].RasterXSize
rows = datasets['B02'].RasterYSize

# Pour chaque date
for date in range(1, nb_dates + 1):
    # Tableau 3D pour stocker toutes les bandes pour cette date
    cube_date = np.zeros((rows, cols, len(bandes)), dtype=np.float32)

    # Pour chaque bande
    for i, bande in enumerate(bandes):
        # Lecture de la bande pour cette date
        data = datasets[bande].GetRasterBand(date).ReadAsArray()
        # Stockage dans le cube
        cube_date[:, :, i] = data

    # Création du fichier de sortie
    driver = gdal.GetDriverByName('GTiff')
    chemin_sortie = f'/home/onyxia/work/results/bretagne_date_{date}.tif'
    out_dataset = driver.Create(chemin_sortie, cols, rows, len(bandes), gdal.GDT_Float32)

    # Copie des informations géoréférencées
    out_dataset.SetGeoTransform(datasets['B02'].GetGeoTransform())
    out_dataset.SetProjection(datasets['B02'].GetProjection())

    # Écriture des bandes
    for i in range(len(bandes)):
        out_band = out_dataset.GetRasterBand(i + 1)
        out_band.WriteArray(cube_date[:, :, i])
        out_band.FlushCache()

    print(f"Image pour la date {date} créée : {chemin_sortie}")

Image pour la date 1 créée : /home/onyxia/work/results/bretagne_date_1.tif
Image pour la date 2 créée : /home/onyxia/work/results/bretagne_date_2.tif
Image pour la date 3 créée : /home/onyxia/work/results/bretagne_date_3.tif
Image pour la date 4 créée : /home/onyxia/work/results/bretagne_date_4.tif
Image pour la date 5 créée : /home/onyxia/work/results/bretagne_date_5.tif


In [15]:
# Liste des bandes
bandes = ['B02', 'B03', 'B04', 'B05', 'B06', 'B07', 'B08', 'B8A', 'B11', 'B12']

# Liste des dates 
dates = ['2024-07-19', '2024-06-07', '2024-04-18', '2024-01-19', '2023-10-08']

# Dictionnaire pour stocker les datasets
datasets = {}

# Chargement des datasets
for bande in bandes:
    chemin = f'/home/onyxia/work/data/projet_eval/bretagne_23-24_{bande}.tif'
    datasets[bande] = gdal.Open(chemin, gdal.GA_ReadOnly)

# Dimensions spatiales
cols = datasets['B02'].RasterXSize
rows = datasets['B02'].RasterYSize

# Pour chaque date (indexé de 1 à 5)
for date_idx in range(1, len(dates) + 1):
    date = dates[date_idx - 1]  # Récupère la date correspondante
    # Tableau 3D pour stocker toutes les bandes pour cette date
    cube_date = np.zeros((rows, cols, len(bandes)), dtype=np.float32)

    # Pour chaque bande
    for i, bande in enumerate(bandes):
        # Lecture de la bande pour cette date
        data = datasets[bande].GetRasterBand(date_idx).ReadAsArray()
        cube_date[:, :, i] = data

    # Création du fichier de sortie (avec la date dans le nom)
    driver = gdal.GetDriverByName('GTiff')
    chemin_sortie = f'/home/onyxia/work/results/bretagne_{date}.tif'
    out_dataset = driver.Create(chemin_sortie, cols, rows, len(bandes), gdal.GDT_Float32)

    # Copie des informations géoréférencées
    out_dataset.SetGeoTransform(datasets['B02'].GetGeoTransform())
    out_dataset.SetProjection(datasets['B02'].GetProjection())

    # Écriture des bandes
    for i in range(len(bandes)):
        out_band = out_dataset.GetRasterBand(i + 1)
        out_band.WriteArray(cube_date[:, :, i])
        out_band.FlushCache()

    print(f"Image pour la date {date} créée : {chemin_sortie}")

Image pour la date 2024-07-19 créée : /home/onyxia/work/results/bretagne_2024-07-19.tif
Image pour la date 2024-06-07 créée : /home/onyxia/work/results/bretagne_2024-06-07.tif
Image pour la date 2024-04-18 créée : /home/onyxia/work/results/bretagne_2024-04-18.tif
Image pour la date 2024-01-19 créée : /home/onyxia/work/results/bretagne_2024-01-19.tif
Image pour la date 2023-10-08 créée : /home/onyxia/work/results/bretagne_2023-10-08.tif


## Analyse des échantillons
### Nombre d'échantillons

In [6]:
import geopandas as gpd
import plotly.express as px
import os
import pandas as pd

# Chemin vers le fichier shapefile
shp_path = '/home/onyxia/work/data/projet_eval/PI_strates_bretagne_32630.shp'

# Charger le fichier
gdf = gpd.read_file(shp_path)

# Compter le nombre de polygones par classe
polygon_counts = gdf['strate'].value_counts().reset_index()
polygon_counts.columns = ['Classe', 'Nombre de polygones']

# Convertir 'Classe' en chaîne de caractères pour éviter les conflits de type
polygon_counts['Classe'] = polygon_counts['Classe'].astype(str)

# Créer un DataFrame avec toutes les classes (1, 2, 3, 4), même si elles sont absentes
all_classes = pd.DataFrame({'Classe': ['1', '2', '3', '4']})

# Fusionner avec les comptes réels
polygon_counts = all_classes.merge(polygon_counts, on='Classe', how='left').fillna(0)

# Dictionnaire de correspondance classe/couleur (personnalisable)
color_map = {
    '1': '#F5DEB3',  # Sol nu (tan, couleur personnalisée)
    '2': '#98FB98',  # Herbe (palegreen, personnalisée)
    '3': '#32CD32',  # Landes (limegreen, personnalisée)
    '4': '#228B22'   # Arbre (darkgreen, personnalisée)
}

# Créer le diagramme avec Plotly
fig_polygons = px.bar(
    polygon_counts,
    x='Classe',
    y='Nombre de polygones',
    color='Classe',
    color_discrete_map=color_map,
    title='Nombre de polygones par classe',
    labels={'Classe': 'Classe', 'Nombre de polygones': 'Nombre de polygones'},
    category_orders={'Classe': ['1', '2', '3', '4']}
)

# Personnaliser l'affichage des classes sur l'axe x
fig_polygons.update_xaxes(
    type='category',  # Forcer l'affichage en catégories discrètes
    tickvals=['1', '2', '3', '4'],  # Valeurs exactes à afficher
    ticktext=['Sol nu', 'Herbe', 'Landes', 'Arbre']   # Libellés à afficher
)

# Enregistrer la figure
output_polygon_path = '/home/onyxia/work/results/figure/diag_baton_nb_poly_by_class.html'
os.makedirs(os.path.dirname(output_polygon_path), exist_ok=True)
fig_polygons.write_html(output_polygon_path)

# Afficher la figure dans le notebook
fig_polygons.show()


La classe 1 (sol nu) est absente des 

In [9]:
# Calculer l'aire de chaque polygone
gdf['Aire'] = gdf.geometry.area

# Calculer l'aire totale par classe
aire_counts = gdf.groupby('strate')['Aire'].sum().reset_index()
aire_counts.columns = ['Classe', 'Aire totale']

# Convertir 'Classe' en chaîne de caractères
aire_counts['Classe'] = aire_counts['Classe'].astype(str)

# Fusionner avec toutes les classes
all_classes = pd.DataFrame({'Classe': ['1', '2', '3', '4']})
aire_counts = all_classes.merge(aire_counts, on='Classe', how='left').fillna(0)

# Créer le diagramme
fig_aires = px.bar(
    aire_counts,
    x='Classe',
    y='Aire totale',
    color='Classe',
    color_discrete_map=color_map,
    title='Aire totale par classe (proxy pour le nombre de pixels)',
    labels={'Classe': 'Classe', 'Aire totale': 'Aire totale (m²)'},
    category_orders={'Classe': ['1', '2', '3', '4']}
)

# Personnaliser l'axe x
fig_aires.update_xaxes(
    type='category',
    tickvals=['1', '2', '3', '4'],
    ticktext=['Sol nu', 'Herbe', 'Landes', 'Arbre']
)

# Enregistrer la figure
output_aire_path = '/home/onyxia/work/results/figure/diag_baton_nb_pix_by_class.html'
fig_aires.write_html(output_aire_path)

# Afficher la figure
fig_aires.show()


In [ ]:
# define parameters
my_folder = "/home/onyxia/work/Projet_teledetection"
in_vector = os.path.join(my_folder, '../data/projet_eval/Pi_strates_bretagne_32630.shp')
ref_image = os.path.join(my_folder, '../data/projet_eval/bretagne_23-24_B02.tif')
out_image = os.path.splitext(in_vector)[0] + '_rasterize.tif'
field_name = 'num'  # field containing the numeric label of the classes

# for those parameters, you know how to get theses information if you had to
sptial_resolution = 0.5
xmin = 748231.0
ymin = 6273800.0
xmax = 751231.0
ymax = 6276800.0

# define command pattern to fill with paremeters
cmd_pattern = ("gdal_rasterize -a {field_name} "
               "-tr {sptial_resolution} {sptial_resolution} "
               "-te {xmin} {ymin} {xmax} {ymax} -ot Byte -of GTiff "
               "{in_vector} {out_image}")

# fill the string with the parameter thanks to format function
cmd = cmd_pattern.format(in_vector=in_vector, xmin=xmin, ymin=ymin, xmax=xmax,
                         ymax=ymax, out_image=out_image, field_name=field_name,
                         sptial_resolution=sptial_resolution)

# execute the command in the terminal
os.system(cmd)



In [16]:
# Chemin vers le fichier shapefile
shp_path = '/home/onyxia/work/data/projet_eval/PI_strates_bretagne_32630.shp'

# Charger le fichier shapefile
gdf = gpd.read_file(shp_path)

# Chemin vers une image Sentinel-2 pour les dimensions
reference_image_path = '/home/onyxia/work/results/bretagne_2023-10-08.tif'

# Ouvrir l'image de référence pour obtenir les dimensions et la transformation
reference_ds = gdal.Open(reference_image_path)
cols = reference_ds.RasterXSize
rows = reference_ds.RasterYSize
transform = reference_ds.GetGeoTransform()
projection = reference_ds.GetProjection()

# Créer un raster vide pour la rasterisation
driver = gdal.GetDriverByName('GTiff')
output_raster_path = '/home/onyxia/work/results/temp_rasterized.tif'
raster_ds = driver.Create(output_raster_path, cols, rows, 1, gdal.GDT_Byte)
raster_ds.SetGeoTransform(transform)
raster_ds.SetProjection(projection)
band = raster_ds.GetRasterBand(1)
band.Fill(0)  # Remplir avec 0 (fond)

# Rasteriser les polygones
layer = ogr.Open(shp_path)
gdal.RasterizeLayer(raster_ds, [1], layer, options=["ATTRIBUTE=strate"])

# Lire le raster rasterisé
raster_array = band.ReadAsArray()

# Compter le nombre de pixels par classe
unique, counts = np.unique(raster_array, return_counts=True)
pixel_counts = dict(zip(unique, counts))

# Exclure la classe 0 (fond)
pixel_counts.pop(0, None)

# Créer un DataFrame pour Plotly
pixel_df = pd.DataFrame({
    'Classe': list(pixel_counts.keys()),
    'Nombre de pixels': list(pixel_counts.values())
})

# Ajouter les classes manquantes (si aucune occurrence)
all_classes = ['1', '2', '3', '4']
for classe in all_classes:
    if int(classe) not in pixel_df['Classe'].values:
        pixel_df = pixel_df.append({'Classe': int(classe), 'Nombre de pixels': 0}, ignore_index=True)

# Trier par classe
pixel_df = pixel_df.sort_values('Classe')

# Dictionnaire de correspondance classe/couleur
color_map = {
    '1': '#F5DEB3',  # Sol nu
    '2': '#98FB98',  # Herbe
    '3': '#32CD32',  # Landes
    '4': '#228B22'   # Arbre
}

# Créer le diagramme avec Plotly
fig_pixels = px.bar(
    pixel_df,
    x='Classe',
    y='Nombre de pixels',
    color='Classe',
    color_discrete_map={str(k): v for k, v in color_map.items()},
    title='Nombre de pixels par classe',
    labels={'Classe': 'Classe', 'Nombre de pixels': 'Nombre de pixels'},
    category_orders={'Classe': ['1', '2', '3', '4']}
)

# Personnaliser l'axe x
fig_pixels.update_xaxes(
    type='category',
    tickvals=['1', '2', '3', '4'],
    ticktext=['1', '2', '3', '4']
)

# Enregistrer la figure
output_pixel_path = '/home/onyxia/work/results/figure/diag_baton_nb_pix_by_class.html'
os.makedirs(os.path.dirname(output_pixel_path), exist_ok=True)
fig_pixels.write_html(output_pixel_path)

# Afficher la figure
fig_pixels.show()


TypeError: in method 'RasterizeLayer', argument 4 of type 'OGRLayerShadow *'

In [ ]:
# Chemins des fichiers
my_folder = "/home/onyxia/work/Projet_teledetection"
in_vector = os.path.join(my_folder, '../data/projet_eval/PI_strates_bretagne_32630.shp')
ref_image = os.path.join(my_folder, '../data/projet_eval/bretagne_23-24_B02.tif')
out_image = os.path.join(my_folder, '../data/projet_eval/PI_strates_bretagne_32630_rasterize.tif')

# Ouvrir l'image de référence pour extraire les paramètres
# La résolution spatiale et l'emprise sont celles obtenues à partir de l'une des images initiales, ici bretagne_23-24_B02.tif
ref_ds = gdal.Open(ref_image)
transform = ref_ds.GetGeoTransform()
xmin = transform[0]
ymin = transform[3] + transform[5] * ref_ds.RasterYSize  # Coin inférieur gauche
xmax = transform[0] + transform[1] * ref_ds.RasterXSize  # Coin supérieur droit
ymax = transform[3]
spatial_resolution = transform[1]  # Résolution en x (supposée égale en y)

print(f"Résolution spatiale : {spatial_resolution}")
print(f"Étendue : xmin={xmin}, ymin={ymin}, xmax={xmax}, ymax={ymax}")


Résolution spatiale : 10.0
Étendue : xmin=432950.0, ymin=5362770.0, xmax=448280.0, ymax=5368890.0


In [18]:
# Définir le champ contenant les labels des classes
field_name = 'strate'  # À adapter selon le nom exact dans ton shapefile

# Commande pour rasteriser
cmd_pattern = (
    f"gdal_rasterize -a {field_name} "
    f"-tr {spatial_resolution} {spatial_resolution} "
    f"-te {xmin} {ymin} {xmax} {ymax} "
    "-ot Byte -of GTiff "
    f"{in_vector} {out_image}"
)

print(f"Commande exécutée : {cmd_pattern}")

# Exécuter la commande
os.system(cmd_pattern)


Commande exécutée : gdal_rasterize -a strate -tr 10.0 10.0 -te 432950.0 5362770.0 448280.0 5368890.0 -ot Byte -of GTiff /home/onyxia/work/Projet_teledetection/../data/projet_eval/PI_strates_bretagne_32630.shp /home/onyxia/work/Projet_teledetection/../data/projet_eval/PI_strates_bretagne_32630_rasterize.tif
0...10...20...30...40...50...60...70...80...90...100 - done.


0

In [19]:
# Ouvrir le raster rasterisé
raster_ds = gdal.Open(out_image)
band = raster_ds.GetRasterBand(1)
raster_array = band.ReadAsArray()

# Compter les pixels par classe
unique, counts = np.unique(raster_array, return_counts=True)
pixel_counts = dict(zip(unique, counts))

# Exclure la classe 0 (fond)
pixel_counts.pop(0, None)

print("Nombre de pixels par classe :", pixel_counts)


Nombre de pixels par classe : {np.uint8(2): np.int64(1022), np.uint8(3): np.int64(1599), np.uint8(4): np.int64(1218)}


In [20]:
# Créer un DataFrame pour Plotly
pixel_df = pd.DataFrame({
    'Classe': list(pixel_counts.keys()),
    'Nombre de pixels': list(pixel_counts.values())
})

# Ajouter les classes manquantes (si aucune occurrence)
all_classes = ['1', '2', '3', '4']
for classe in all_classes:
    if int(classe) not in pixel_df['Classe'].values:
        pixel_df = pixel_df.append({'Classe': int(classe), 'Nombre de pixels': 0}, ignore_index=True)

# Trier par classe
pixel_df = pixel_df.sort_values('Classe')

# Dictionnaire de couleurs
color_map = {
    '1': '#F5DEB3',  # Sol nu
    '2': '#98FB98',  # Herbe
    '3': '#32CD32',  # Landes
    '4': '#228B22'   # Arbre
}

# Créer le diagramme
fig_pixels = px.bar(
    pixel_df,
    x='Classe',
    y='Nombre de pixels',
    color='Classe',
    color_discrete_map={str(k): v for k, v in color_map.items()},
    title='Nombre de pixels par classe',
    labels={'Classe': 'Classe', 'Nombre de pixels': 'Nombre de pixels'},
    category_orders={'Classe': ['1', '2', '3', '4']}
)

# Personnaliser l'axe x
fig_pixels.update_xaxes(
    type='category',
    tickvals=['1', '2', '3', '4'],
    ticktext=['1', '2', '3', '4']
)

# Enregistrer la figure
output_pixel_path = '/home/onyxia/work/results/figure/diag_baton_nb_pix_by_class.html'
os.makedirs(os.path.dirname(output_pixel_path), exist_ok=True)
fig_pixels.write_html(output_pixel_path)

# Afficher la figure
fig_pixels.show()


AttributeError: 'DataFrame' object has no attribute 'append'